pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/BulkProcessor.java

```diff
@@ -0,0 +1,335 @@
+        this.client = client;
+        this.lock = new ReentrantLock();
+        this.bulkActions = config.getBulkActions();
+        this.bulkSize = config.getBulkSizeInMb() * 1024 * 1024;
+        this.internalExecutorService = Executors.newFixedThreadPool(Math.max(1, config.getBulkConcurrentRequests()),
+                new DefaultThreadFactory("elastic-bulk-executor"));
+        this.bulkRequestHandler = new BulkRequestHandler(new RandomExponentialRetry(config.getMaxRetryTimeInSec()),
+                config.getBulkConcurrentRequests(), listener);
+
+        ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(
+                new DefaultThreadFactory("elastic-flusher"));
+        if (config.getBulkFlushIntervalInMs() > 0) {
+            futureFlushTask = executor.scheduleWithFixedDelay(new Flush(),
+                    config.getBulkFlushIntervalInMs(),
+                    config.getBulkFlushIntervalInMs(),
+                    TimeUnit.MILLISECONDS);
+
+        }
+
+    }
+
+    protected void ensureOpen() {
+        if (this.closed) {
+            throw new IllegalStateException("bulk process already closed");
+        }
+    }
+
+    private BulkRequest createBulkRequestAndResetPendingOps() {
+        final BulkRequest bulkRequest = new BulkRequest.Builder()
+                .operations(new ArrayList<>(pendingOperations))
+                .build();
+        this.pendingOperations.clear();
+        return bulkRequest;
+    }
+
+    private void execute() {
+        this.bulkRequest = createBulkRequestAndResetPendingOps();
+        long executionId = this.executionIdGen.incrementAndGet();
+        this.execute(bulkRequest, executionId);
+    }
+
+    private boolean isOverTheLimit() {
+        if (this.bulkActions != -1 && pendingOperations.size() >= this.bulkActions) {
+            return true;
+        } else {
+            return this.bulkSize != -1L &&
+                    pendingOperations.stream().mapToLong(op -> op.getEstimatedSizeInBytes()).sum() >= this.bulkSize;
+        }
+    }
+
+    public void flush() {
+        this.lock.lock();
+
+        try {
+            this.ensureOpen();
+            if (pendingOperations.size() > 0) {
+                this.execute();
+            }
+        } finally {
+            this.lock.unlock();
+        }
+
+    }
+
+    private void execute(BulkRequest bulkRequest, long executionId) {
+        this.bulkRequestHandler.execute(bulkRequest, executionId);
+    }
+
+    private Pair<BulkRequest, Long> newBulkRequestIfNeeded() {
+        this.ensureOpen();
+        if (!this.isOverTheLimit()) {
+            return null;
+        } else {
+            final BulkRequest bulkRequest = createBulkRequestAndResetPendingOps();
+            return Pair.of(bulkRequest, this.executionIdGen.incrementAndGet());
+        }
+    }
+
+    public void add(BulkOperationWithId bulkOperation) {
+        Pair<BulkRequest, Long> bulkRequestToExecute;
+        this.lock.lock();
```

Comment:
this is a very coarse grained lock !
basically even if you have multiple threads we are not going to do anything in parallel.

in order to retain ordering we are not going to set concurrency level greater that 1, so for DataStax/CDC usecases this may be good.

but please take into account that we have this lock both for managing the queue and for doing IO

Reply:
the concurrency is only for IO and it's handled with a Semaphore https://github.com/datastax/pulsar/blob/058068943b5a507d57ebe97f11718931e95b40b7/pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/BulkProcessor.java#L274-L322


the lock here is needed to compute atomical operation while reading and clearing the pendingOperations list. This can be done from the 'main' thread (which calls `add`) and the periodic flush task

however I will refactor it to make it simpler